### ISSUE TITLE
SNOW-375565: How to `fetchAsString` but keep null values?

### ISSUE THREAD

**Issue by datbth** [snowflakedb/snowflake-connector-nodejs#176]
Please answer these questions before submitting your issue. Thanks!

1. What version of NodeJS are you using (`node --version` and `npm --version`)?  
Node: v14.17.0  
Npm: 6.14.14

2. What operating system and processor architecture are you using?
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```

3. What are the component versions in the environment (`npm list`)?
I'm using snowflake-sdk@1.6.1

4. What did you do?
Run a SQL with fetchAsString option:
```javascript
connection.execute({
  sqlText: 'select null::float',
  fetchAsString: ['NUMBER'],
  complete (err, stmt, rows) {
    console.log(rows[0][0] === null);
    console.log(JSON.stringify(rows[0][0]));
  }
```

5. What did you expect to see?
Console logging `true` and `null`

6. What did you see instead?
Console logging `false` and `"null"`

7. Add this to get standard output.
```
{"message":"[10:51:52.5152 AM]: Contacting SF: /session/v1/login-request?requestId=REQUEST_ID&warehouse=TEST&databaseName=TEST, (1/5)","level":"DEBUG"}
{"message":"[10:51:52.5152 AM]: SNOWFLAKE_URL/session/v1/login-request?requestId=REQUEST_ID&warehouse=TEST&databaseName=TEST","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: Reading OCSP cache file. /home/datbth/.cache/snowflake/ocsp_response_cache.json","level":"DEBUG"}
{"message":"[10:51:52.5152 AM]: Returning OCSP status for certificate 075EBC4CCBC3E05D031EF502C6175674 from cache","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: Returning OCSP status for certificate 067F94578587E8AC77DEB253325BBC998B560D from cache","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: Returning OCSP status for certificate 067F944A2A27CDF3FAC2AE2B01F908EEB9C4C6 from cache","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: Returning OCSP status for certificate A70E4A4C3482B77F from cache","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: socket reused = false","level":"TRACE"}
{"message":"[10:51:52.5152 AM]: OCSP validation succeeded for holistics.ap-southeast-2.snowflakecomputing.com","level":"TRACE"}
{"message":"[10:51:53.5153 AM]: SNOWFLAKE_URL/queries/v1/query-request?requestId=ab0b5126-6fe0-4762-94b0-96587c53b8ef","level":"TRACE"}
{"message":"[10:51:53.5153 AM]: socket reused = true","level":"TRACE"}
```

-----

I would like to fetch all values as raw Strings, so that I don't have to worry about any lossy/distorted values due to Javascript processing. But currently, with `fetchAsString` option, `null` values are being turned into `'NULL'` (NULL_UPPERCASE string). I could process through the non-text columns and revert `'NULL'` values into `null`s, but that would be very inefficient as the values have to be processed unnecessarily multiple times.  
Thus, may I ask for the reasons behind the decision of this behavior of turning `null` into `'NULL'`? And whether there are any better setup or workaround for my case?  
Thank you.

**Comment by barooo** [comment]
Is there a resolution here?  This seems to still be an issue.

**Comment by sfc-gh-dszmolka** [comment]
hi, thank you for submitting this issue. using latest (`v1.6.18`) of the driver for testing.
1. executing `select null, SYSTEM$TYPEOF(null) as typeof` **with or without** `fetchAsString` returns
```
{ NULL: null, TYPEOF: 'NULL[LOB]' }
```
2. executing `select null::float, SYSTEM$TYPEOF(null::float) as typeof` **with** `fetchAsString: ['NUMBER']` returns:
```
{ 'NULL::FLOAT': 'NULL', TYPEOF: 'FLOAT[DOUBLE]' }
```

per the Snowflake documentation for [Fetching Data Types as Strings](https://docs.snowflake.com/en/user-guide/nodejs-driver-use#fetching-data-types-as-strings)
> When connection.execute() is called, the fetchAsString option can be set to force all numbers or dates to be returned as strings.

as i see it, this is what is happening, forcing the number to be returned as string. (`'NULL'`) so at first glance this seems to be the expected and documented behaviour: fetches number as string. When `fetchAsString` is omitted, the `null` seems to be returned as `null` 

what happens in your case when you leave `fetchAsString: ['NUMBER']` but do not transform the `null` into `float` so it would be fetched as string ?
also does perhaps the Snowflake function [TO_VARCHAR](https://docs.snowflake.com/en/sql-reference/functions/to_char) provide any workaround? It should cast the column's value to string, and if it's `NULL` then leaves as `NULL`. Since the transformation is happening on server-side, you would not need `fetchAsString`  in this case.

**Comment by datbth** [comment]
Hi @sfc-gh-dszmolka, thank you for the response.

> what happens in your case when you leave fetchAsString: ['NUMBER'] but do not transform the null into float so it would be fetched as string ?

`select null::float` is only an example to demonstrate the issue.  
I'm seeing the same issue when querying NULL values from a float column of a table. E.g.
```sql
CREATE TABLE a (x float);
INSERT INTO a VALUES (null);
SELECT x FROM a;
```

> also does perhaps the Snowflake function [TO_VARCHAR](https://docs.snowflake.com/en/sql-reference/functions/to_char) provide any workaround?

* In my case, I'm developing a BI (Business Intelligence) tool that does not have control over the user's SQLs. But even in other post-processing use cases, especially in data processing pipelines, I think this would still be quite an **inconvenient** workaround and it's limiting the connector's usability in data processing pipelines.
* More importantly, this issue is about the **inefficiency** of handling NULL values. I believe using something like `TO_VARCHAR` would still yield some unnecessary processing overhead.

**Comment by sfc-gh-ext-simba-ti** [comment]
Triaged. Keeping this with me as FR. 
I've reviewed Simba case 00345547, no access to SNOW-375565 atm. 
Last notes in the case was that this was reviewed with Snowflake team on July 2021, and another note in Nov 2022 for this to be prioritized by PM. 

Discuss with Sadia/Anurag to revisit this in 2023.Q2.

**Comment by sfc-gh-dszmolka** [comment]
thank you for the review and summary and sharing the plans Toar ! since you mentioned you don't have access to the jira, i'm attaching it here for completeness. there's not much new information in it, but to have a full picture.
[SNOW-375565.pdf](https://github.com/snowflakedb/snowflake-connector-nodejs/files/11176582/SNOW-375565.pdf)

**Comment by sfc-gh-dszmolka** [comment]
thank you for sharing your approach of how to work around this issue ! sadly I cannot answer the question why `NULL_UPPERCASE` is defined this way, it looks like per the blame, it was always like this. Doesn't mean needs to stay like this though. I'll keep this ticket posted once we have progress on it.

**Comment by sfc-gh-dszmolka** [comment]
PR in review at https://github.com/snowflakedb/snowflake-connector-nodejs/pull/831

**Comment by sfc-gh-dszmolka** [comment]
PR merged and fix will be part of the next release cycle, probably end of May 2024

**Comment by sfc-gh-dszmolka** [comment]
released with May 2024 release cycle, version 1.11.0

thank you for your patience here !
